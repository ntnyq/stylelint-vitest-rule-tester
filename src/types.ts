import type { Config, LinterOptions, LinterResult } from 'stylelint'

/**
 * default filename when not provided
 */
export interface DefaultFilenames {
  css: string
  less: string
  postcss: string
  sass: string
  scss: string
  styl: string
  stylus: string
  [key: string]: string
}

/**
 * rule to test
 */
export type RuleModule = any

/**
 * Rule Tester
 */
export interface RuleTester {
  /**
   * run a single test case
   */
  each: (arg: TestCase) => Promise<TestExecutionResult>

  /**
   * run a single valid test case
   */
  valid: (arg: ValidTestCase) => Promise<TestExecutionResult>

  /**
   * run a single invalid test case
   */
  invalid: (arg: InvalidTestCase) => Promise<TestExecutionResult>

  /**
   * run multiple test cases
   */
  run: (options: TestCasesOptions) => void
}
export interface RuleTesterBehaviorOptions {
  /**
   * the number of times to recursively apply the rule
   *
   * @default 10
   */
  recursive?: false | number

  /**
   * run verification after applying the fix
   *
   * @default true
   */
  verifyAfterFix?: boolean

  /**
   * verify that fix allways changes the code
   *
   * @default true
   */
  verifyFixChanges?: boolean
}
export interface RuleTesterInitOptions extends RuleTesterBehaviorOptions, StylelintOptions {
  /**
   * the rule to test
   */
  rule?: RuleModule

  /**
   * the name of the rule to test
   */
  name?: string

  /**
   * the default filenames to be used for tests
   */
  defaultFileNames?: Partial<DefaultFilenames>
}

/**
 * Test
 */
export interface TestCasesOptions {
  /**
   * invalid cases
   */
  invalid?: (string | InvalidTestCase)[]

  /**
   * valid cases
   */
  valid?: (string | ValidTestCase)[]

  /**
   * callback to be called after each test case
   */
  onResult?: (testCase: NormalizedTestCase, result: LinterResult) => Promise<void> | void
}
export interface TestExecutionResult extends LinterResult {
  /**
   * whether the code was fixed
   */
  fixed?: boolean

  /**
   * if the rule fixes in multiple steps, the result of each step is present here
   */
  steps?: LinterResult[]
}

/**
 * Stylelint
 */
export interface LintResultDeprecation {
  text: string
  reference?: string
}
export interface LintResultInvalidOptionWarning {
  text: string
}
export type LintResultMessage =
  | LintResultDeprecation
  | LintResultInvalidOptionWarning
  | LintResultParseError
  | LintResultWarning
export interface LintResultParseError {
  column: number
  line: number
  endColumn?: number
  endLine?: number
  // postcss node
  node: any
  plugin: string
  stylelintType: 'parseError'
  text: string
  type: 'warning'
}
export interface LintResultWarning {
  column: number
  line: number
  rule: string
  severity: 'error' | 'warning'
  text: string
  endColumn?: number
  endLine?: number
  stylelintType?: 'deprecation' | 'invalidOption' | 'parseError'
  url?: string
}
export interface StylelintOptions {
  /**
   * stylelint config
   */
  stylelintConfig?: Config

  /**
   * linter options for `stylelint.lint(options)`
   */
  linterOptions?: LinterOptions
}

/**
 * Test case
 */
export type InvalidTestCase = string | InvalidTestCaseBase
export interface InvalidTestCaseBase extends ValidTestCaseBase {
  warnings?: number | (string | LintResultWarning)[] | ((warnings: LintResultWarning[]) => void)

  deprecations?:
    | number
    | (string | LintResultDeprecation)[]
    | ((deprecations: LintResultDeprecation[]) => void)

  invalidOptionWarnings?:
    | number
    | (string | LintResultInvalidOptionWarning)[]
    | ((invalidOptionWarnings: LintResultInvalidOptionWarning[]) => void)

  parseErrors?:
    | number
    | (string | LintResultParseError)[]
    | ((parseErrors: LintResultParseError[]) => void)

  /**
   * Assert if output is expected.
   * Pass `null` to assert that the output is the same as the input.
   */
  output?: string | ((output: string, input: string) => void) | null
}
export interface NormalizedTestCase extends InvalidTestCaseBase {
  code: string
  type: 'invalid' | 'valid'
}
export type TestCase = InvalidTestCase | ValidTestCase
export type ValidTestCase = string | ValidTestCaseBase
export interface ValidTestCaseBase extends RuleTesterBehaviorOptions, StylelintOptions {
  /**
   * test case name
   */
  name?: string

  /**
   * test case description
   */
  description?: string

  /**
   * code to test
   */
  code: string

  /**
   * test case filename
   */
  filename?: string

  /**
   * only run this test case
   */
  only?: boolean

  /**
   * skip this test case
   */
  skip?: boolean

  /**
   * rule options
   */
  ruleOptions?: any

  /**
   * lint result
   */
  onResult?: (result: LinterResult) => void
}
