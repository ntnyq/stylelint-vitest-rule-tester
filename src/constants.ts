import type { DefaultFilenames } from './types'

export const DEFAULT_RULE_NAME = 'rule-to-test'
export const DEFAULT_FILE_NAME = 'file.css'

export const DEFAULT_RULE_OPTIONS = true

export const DEFAULT_FILE_NAMES: DefaultFilenames = {
  css: 'file.css',
  less: 'file.less',
  postcss: 'file.css',
  sass: 'file.sass',
  scss: 'file.scss',
  styl: 'file.styl',
  stylus: 'file.stylus',
}
