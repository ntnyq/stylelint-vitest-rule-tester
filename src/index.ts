export * from './run'
export * from './types'
export * from './utils'
export * from './tester'
