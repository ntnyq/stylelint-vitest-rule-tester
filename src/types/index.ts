export * from './case'
export * from './file'
export * from './test'
export * from './tester'
export * from './stylelint'
