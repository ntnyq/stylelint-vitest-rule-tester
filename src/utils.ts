import { isString, unindent } from '@ntnyq/utils'
import {
  DEFAULT_FILE_NAME,
  DEFAULT_RULE_NAME,
  DEFAULT_RULE_OPTIONS,
} from './constants'
import type {
  DefaultFilenames,
  LintResultMessage,
  NormalizedTestCase,
  RuleTesterInitOptions,
  TestCase,
  TestExecutionResult,
} from './types'

/**
 * Get rule name
 * @param options - rule tester options
 * @returns rule name
 */
export function getRuleName(options: RuleTesterInitOptions) {
  return options.name || options.rule?.ruleName || DEFAULT_RULE_NAME
}

export function normalizeRuleOptions(
  testCase: NormalizedTestCase,
  options: RuleTesterInitOptions,
) {
  const { rule = {} } = options
  const { url } = rule?.meta || {}

  const resolvedOptions = testCase.ruleOptions || options.ruleOptions

  if (resolvedOptions) {
    if (Array.isArray(resolvedOptions)) {
      return resolvedOptions.length === 1
        ? [resolvedOptions[0], { url }]
        : resolvedOptions
    } else {
      return [resolvedOptions, { url }]
    }
  }

  return url ? [DEFAULT_RULE_OPTIONS, { url }] : DEFAULT_RULE_OPTIONS
}

/**
 * Check if given test case is invalid
 *
 * @param c - normalized test case
 * @returns true if given case is invalid
 */
function isInvalidTestCase(c: NormalizedTestCase): boolean {
  return (
    !!c.warnings
    || !!c.deprecations
    || !!c.parseErrors
    || !!c.invalidOptionWarnings
    || !!c.output
  )
}

export function normalizeTestCase(
  c: TestCase,
  defaultFilenames: Partial<DefaultFilenames>,
  type?: 'valid' | 'invalid',
): NormalizedTestCase {
  const obj = isString(c) ? { code: c } : { ...c }
  const normalized = obj as NormalizedTestCase

  normalized.type ||=
    type || (isInvalidTestCase(normalized) ? 'invalid' : 'valid')

  normalized.filename ||= defaultFilenames.css || DEFAULT_FILE_NAME

  return normalized
}

/**
 * Normalize test case message
 *
 * @param message - message string or lint result
 * @returns normalized message
 */
export function normalizeCaseMessage(
  message: string | LintResultMessage,
): Partial<LintResultMessage> {
  if (isString(message)) {
    return {
      text: message,
    }
  }
  const clone = { ...message }

  return clone as Partial<LintResultMessage>
}

/**
 * Safe access from linter result
 */
export function normalizeLinterResult(linterResult: TestExecutionResult) {
  const { fixed = false } = linterResult
  const {
    errored = false,
    warnings = [],
    parseErrors = [],
    deprecations = [],
    invalidOptionWarnings = [],
  } = linterResult.results[0]

  return {
    fixed,
    errored,
    warnings,
    parseErrors,
    deprecations,
    invalidOptionWarnings,
  }
}

export { unindent as $, unindent }
