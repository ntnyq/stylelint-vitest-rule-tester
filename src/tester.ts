import { isFunction, isNull, isNumber, isUndefined, toArray } from '@ntnyq/utils'
import stylelint from 'stylelint'
import { describe, expect, it } from 'vitest'
import { DEFAULT_FILE_NAMES } from './constants'
import {
  getRuleName,
  normalizeCaseMessage,
  normalizeRuleOptions,
  normalizeTestCase,
  useLinterResult,
} from './utils'
import type { Config as LinterConfig, LinterOptions } from 'stylelint'
import type {
  DefaultFilenames,
  InvalidTestCase,
  LintResultDeprecation,
  LintResultInvalidOptionWarning,
  LintResultParseError,
  LintResultWarning,
  NormalizedTestCase,
  RuleTester,
  RuleTesterInitOptions,
  TestCase,
  TestCasesOptions,
  TestExecutionResult,
  ValidTestCase,
} from './types'

function verifyLintResultMessages({
  type,
  testCase,
  messages,
}: {
  type: 'warnings'
  testCase: NormalizedTestCase
  messages: LintResultWarning[]
}): void
function verifyLintResultMessages({
  type,
  testCase,
  messages,
}: {
  type: 'parseErrors'
  testCase: NormalizedTestCase
  messages: LintResultParseError[]
}): void
function verifyLintResultMessages({
  type,
  testCase,
  messages,
}: {
  type: 'deprecations'
  testCase: NormalizedTestCase
  messages: LintResultDeprecation[]
}): void
function verifyLintResultMessages({
  type,
  testCase,
  messages,
}: {
  type: 'invalidOptionWarnings'
  testCase: NormalizedTestCase
  messages: LintResultInvalidOptionWarning[]
}): void
function verifyLintResultMessages({
  type,
  testCase,
  messages,
}: {
  type: 'warnings' | 'parseErrors' | 'deprecations' | 'invalidOptionWarnings'
  testCase: NormalizedTestCase
  messages: any[]
  // | LintResultWarning[]
  // | LintResultParseError[]
  // | LintResultDeprecation[]
  // | LintResultInvalidOptionWarning[]
}) {
  if (!testCase[type]) {
    return
  }

  if (isFunction(testCase[type])) {
    testCase[type]?.(messages)
  } else if (isNumber(testCase[type])) {
    expect.soft(messages.length, `number of ${type}`).toBe(testCase[type])
  } else {
    const testCaseMessages = toArray(testCase[type]).map(message => normalizeCaseMessage(message))

    expect(testCaseMessages.length, `number of ${type}`).toBe(testCase[type].length)

    testCaseMessages.forEach((expected, idx) => {
      expect.soft(messages[idx], `object of ${type}-${idx}`).toMatchObject(expected)
    })
  }
}

export function createRuleTester(options: RuleTesterInitOptions): RuleTester {
  const defaultFilenames: Partial<DefaultFilenames> = {
    ...DEFAULT_FILE_NAMES,
    ...options.defaultFileNames,
  }
  const ruleName = getRuleName(options)

  async function each(c: TestCase) {
    const testCase = normalizeTestCase(c, defaultFilenames)

    const {
      recursive = 10,
      verifyAfterFix = true,
      // verifyFixChanges = true,
    } = {
      ...options,
      ...testCase,
    }

    const linterConfig: LinterConfig = {
      ...options.stylelintConfig,
      rules: {
        [ruleName]: normalizeRuleOptions(testCase, options),
      },
    }

    const lintOptions: LinterOptions = {
      ...options.linterOptions,
      config: linterConfig,
      code: testCase.code,
      codeFilename: testCase.filename,
      fix: false,
    }

    const linterResult = await stylelint.lint(lintOptions)
    const {
      // errored,
      warnings,
      // parseErrors,
      // deprecations,
      // invalidOptionWarnings,
    } = useLinterResult(linterResult)

    verifyLintResultMessages({
      type: 'warnings',
      testCase,
      messages: warnings,
    })
    // verifyLintResultMessages({
    //   type: 'parseErrors',
    //   testCase,
    //   messages: parseErrors,
    // })
    // verifyLintResultMessages({
    //   type: 'deprecations',
    //   testCase,
    //   messages: deprecations,
    // })
    // verifyLintResultMessages({
    //   type: 'invalidOptionWarnings',
    //   testCase,
    //   messages: invalidOptionWarnings,
    // })

    async function fix(code: string) {
      const linterResult = await stylelint.lint({
        ...lintOptions,
        code,
        fix: true,
      })

      // check only invalid should change after fix
      // maybe check linterResult.report is an empty string
      // if (linterResult.code === code && verifyFixChanges) {
      //   throw new Error(`Expected code to be changed after fix, but it's the same as the input.`)
      // }
      return {
        ...linterResult,
        fixed: linterResult.code !== code, // check if the code is changed
      }
    }

    const fixedLinterResult = await fix(testCase.code)

    const result: TestExecutionResult = {
      ...fixedLinterResult,
      steps: [fixedLinterResult],
    }

    // should recursive fix
    if (result.fixed && recursive !== false) {
      let r = recursive

      for (r = recursive; r >= 0; r--) {
        const step = await fix(result.code!)

        result.steps?.push(step)
        result.code = step.code

        if (!step.fixed) {
          break
        }
      }

      if (r === 0) {
        throw new Error(
          `Fix recursion limit exceeded, possibly the fix is not stable. Last output:\n-------\n${result.code}\n-------`,
        )
      }
    }

    // expected fixed
    if (!isUndefined(testCase.output)) {
      if (isNull(testCase.output)) {
        // null means output should be the same as the input
        expect(result.code, 'output').toBe(testCase.code)
      } else if (isFunction(testCase.output)) {
        // custom assertion
        testCase.output(result.code || '', testCase.code)
      } else {
        expect(result.code, 'output').toBe(testCase.output)
      }
    }

    if (
      testCase.type === 'invalid' &&
      isUndefined(testCase.output) &&
      isUndefined(testCase.warnings)
      // && isUndefined(testCase.parseErrors) &&
      // isUndefined(testCase.deprecations) &&
      // isUndefined(testCase.invalidOptionWarnings)
    ) {
      throw new Error(
        `Invalid test case must have either 'output', 'warnings', 'parseErrors', 'deprecations', or 'invalidOptionWarnings' property.`,
      )
    }

    if (result.fixed && verifyAfterFix) {
      const linterResult = await stylelint.lint({
        ...lintOptions,
        code: result.code,
        fix: false,
      })

      expect.soft(useLinterResult(linterResult).warnings, 'no warnings after fix').toEqual([])
    }

    testCase.onResult?.(result)

    return result
  }

  async function valid(arg: ValidTestCase | string) {
    const linterResult = await each(arg)
    const result = useLinterResult(linterResult)

    expect.soft(result.fixed, 'no need to fix for valid cases').toBeFalsy()
    expect.soft(result.warnings, 'no warnings on valid cases').toEqual([])
    // expect.soft(result.deprecations, 'no deprecations on valid cases').toEqual([])
    // expect.soft(result.parseErrors, 'no parseErrors on valid cases').toEqual([])
    // expect.soft(result.invalidOptionWarnings, 'no invalidOptionWarnings on valid cases').toEqual([])

    return linterResult
  }

  async function invalid(arg: InvalidTestCase | string) {
    const linterResult = await each(arg)
    const result = useLinterResult(linterResult)

    expect.soft(result.warnings, 'expect warnings').not.toEqual([])
    // expect.soft(result.deprecations, 'expect deprecations').not.toEqual([])
    // expect.soft(result.parseErrors, 'expect parseErrors').not.toEqual([])
    // expect.soft(result.invalidOptionWarnings, 'expect invalidOptionWarnings').not.toEqual([])

    return linterResult
  }

  function run(cases: TestCasesOptions) {
    describe(ruleName, () => {
      if (cases.valid?.length) {
        describe('valid', () => {
          cases.valid!.forEach((c, index) => {
            const testCase = normalizeTestCase(c, defaultFilenames, 'valid')
            let run: typeof it | typeof it.only = it

            if (testCase.only) {
              run = it.only
            }
            if (testCase.skip) {
              run = it.skip
            }

            run(`Valid #${index}: ${testCase.description || testCase.code}`, async () => {
              const result = await valid(testCase)
              await cases?.onResult?.(testCase, result)
            })
          })
        })
      }

      if (cases.invalid?.length) {
        describe('invalid', () => {
          cases.invalid!.forEach((c, index) => {
            const testCase = normalizeTestCase(c, defaultFilenames, 'invalid')
            let run: typeof it | typeof it.only = it

            if (testCase.only) {
              run = it.only
            }
            if (testCase.skip) {
              run = it.skip
            }

            run(`Invalid #${index}: ${testCase.description || testCase.code}`, async () => {
              const result = await invalid(testCase)
              await cases?.onResult?.(testCase, result)
            })
          })
        })
      }
    })
  }

  return {
    each,
    valid,
    invalid,
    run,
  }
}
