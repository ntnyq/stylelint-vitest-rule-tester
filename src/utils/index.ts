import { unindent } from '@ntnyq/utils'

export * from './getRuleName'
export * from './normalizeTestCase'
export * from './normalizeRuleOptions'
export * from './resolveLinterOptions'
export * from './normalizeCaseMessage'
export * from './normalizeLinterResult'
export * from './verifyLintResultMessages'

export { unindent as $, unindent }
