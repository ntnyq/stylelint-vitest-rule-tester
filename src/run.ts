import { createRuleTester } from './tester'
import type {
  RuleModule,
  RuleTesterInitOptions,
  TestCasesOptions,
} from './types'

/**
 * Shortcut to run test cases for a rule
 */
export function run(options: TestCasesOptions & RuleTesterInitOptions) {
  const tester = createRuleTester(options)
  return tester.run(options)
}

/**
 * Shortcut to run test cases for a rule in classic style
 */
export function runClassic(
  ruleName: string,
  rule: RuleModule,
  cases: TestCasesOptions,
  options?: RuleTesterInitOptions,
) {
  const tester = createRuleTester({
    rule,
    name: ruleName,
    ...options,
  })
  return tester.run(cases)
}
