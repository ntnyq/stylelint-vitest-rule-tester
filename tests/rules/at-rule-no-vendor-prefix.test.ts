/**
 * @see {@link https://github.com/stylelint/stylelint/tree/main/lib/rules/at-rule-no-vendor-prefix}
 */

import { unindent as $ } from '@ntnyq/utils'
import { expect } from 'vitest'
import { run } from '../../src'

/**
 * default
 */
run({
  name: 'at-rule-no-vendor-prefix',
  valid: [
    `@keyframes { 0% { top: 0; } }`,
    {
      code: `@viewport { orientation: landscape; }`,
      onResult(result) {
        expect(result).toMatchSnapshot()
      },
    },
  ],
  invalid: [
    {
      code: $`
        @-webkit-keyframes { 0% { top: 0; } }
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`"@keyframes { 0% { top: 0; } }"`)
      },
      warnings(warnings) {
        expect(warnings).toMatchInlineSnapshot(`
          [
            {
              "column": 1,
              "endColumn": 19,
              "endLine": 1,
              "fix": undefined,
              "line": 1,
              "rule": "at-rule-no-vendor-prefix",
              "severity": "error",
              "text": "Unexpected vendor-prefixed at-rule "@-webkit-keyframes" (at-rule-no-vendor-prefix)",
              "url": undefined,
            },
          ]
        `)
      },
    },
    {
      code: $`
        @-wEbKiT-kEyFrAmEs { 0% { top: 0; } }
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`"@kEyFrAmEs { 0% { top: 0; } }"`)
      },
      warnings(warnings) {
        expect(warnings).toMatchInlineSnapshot(`
          [
            {
              "column": 1,
              "endColumn": 19,
              "endLine": 1,
              "fix": undefined,
              "line": 1,
              "rule": "at-rule-no-vendor-prefix",
              "severity": "error",
              "text": "Unexpected vendor-prefixed at-rule "@-wEbKiT-kEyFrAmEs" (at-rule-no-vendor-prefix)",
              "url": undefined,
            },
          ]
        `)
      },
    },
    {
      code: $`
        @-WEBKIT-KEYFRAMES { 0% { top: 0; } }
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`"@KEYFRAMES { 0% { top: 0; } }"`)
      },
      warnings(warnings) {
        expect(warnings).toMatchInlineSnapshot(`
          [
            {
              "column": 1,
              "endColumn": 19,
              "endLine": 1,
              "fix": undefined,
              "line": 1,
              "rule": "at-rule-no-vendor-prefix",
              "severity": "error",
              "text": "Unexpected vendor-prefixed at-rule "@-WEBKIT-KEYFRAMES" (at-rule-no-vendor-prefix)",
              "url": undefined,
            },
          ]
        `)
      },
    },
    {
      code: $`
        @-moz-keyframes { 0% { top: 0; } }
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(`"@keyframes { 0% { top: 0; } }"`)
      },
      warnings(warnings) {
        expect(warnings).toMatchInlineSnapshot(`
          [
            {
              "column": 1,
              "endColumn": 16,
              "endLine": 1,
              "fix": undefined,
              "line": 1,
              "rule": "at-rule-no-vendor-prefix",
              "severity": "error",
              "text": "Unexpected vendor-prefixed at-rule "@-moz-keyframes" (at-rule-no-vendor-prefix)",
              "url": undefined,
            },
          ]
        `)
      },
    },
    {
      code: $`
        @-ms-viewport { orientation: landscape; }
      `,
      output(output) {
        expect(output).toMatchInlineSnapshot(
          `"@viewport { orientation: landscape; }"`,
        )
      },
      warnings(warnings) {
        expect(warnings).toMatchInlineSnapshot(`
          [
            {
              "column": 1,
              "endColumn": 14,
              "endLine": 1,
              "fix": undefined,
              "line": 1,
              "rule": "at-rule-no-vendor-prefix",
              "severity": "error",
              "text": "Unexpected vendor-prefixed at-rule "@-ms-viewport" (at-rule-no-vendor-prefix)",
              "url": undefined,
            },
          ]
        `)
      },
      onResult(result) {
        expect(result).toMatchSnapshot()
      },
    },
  ],
})
