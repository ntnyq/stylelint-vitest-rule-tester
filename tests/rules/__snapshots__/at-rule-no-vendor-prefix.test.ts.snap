// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`at-rule-no-vendor-prefix > invalid > Invalid #4: @-ms-viewport { orientation: landscape; } 3`] = `
{
  "code": "@viewport { orientation: landscape; }",
  "errored": false,
  "fixed": true,
  "output": "@viewport { orientation: landscape; }",
  "reportedDisables": [],
  "results": [
    {
      "deprecations": [],
      "errored": false,
      "ignored": undefined,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [],
    },
  ],
  "ruleMetadata": {
    "at-rule-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix",
    },
  },
  "steps": [
    {
      "code": "@viewport { orientation: landscape; }",
      "errored": false,
      "fixed": true,
      "output": "@viewport { orientation: landscape; }",
      "reportedDisables": [],
      "results": [
        {
          "deprecations": [],
          "errored": false,
          "ignored": undefined,
          "invalidOptionWarnings": [],
          "parseErrors": [],
          "warnings": [],
        },
      ],
      "ruleMetadata": {
        "at-rule-no-vendor-prefix": {
          "fixable": true,
          "url": "https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix",
        },
      },
    },
    {
      "code": "@viewport { orientation: landscape; }",
      "errored": false,
      "fixed": false,
      "output": "@viewport { orientation: landscape; }",
      "reportedDisables": [],
      "results": [
        {
          "deprecations": [],
          "errored": false,
          "ignored": undefined,
          "invalidOptionWarnings": [],
          "parseErrors": [],
          "warnings": [],
        },
      ],
      "ruleMetadata": {
        "at-rule-no-vendor-prefix": {
          "fixable": true,
          "url": "https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix",
        },
      },
    },
  ],
}
`;

exports[`at-rule-no-vendor-prefix > valid > Valid #1: @viewport { orientation: landscape; } 1`] = `
{
  "code": "@viewport { orientation: landscape; }",
  "errored": false,
  "fixed": false,
  "output": "@viewport { orientation: landscape; }",
  "reportedDisables": [],
  "results": [
    {
      "deprecations": [],
      "errored": false,
      "ignored": undefined,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [],
    },
  ],
  "ruleMetadata": {
    "at-rule-no-vendor-prefix": {
      "fixable": true,
      "url": "https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix",
    },
  },
  "steps": [
    {
      "code": "@viewport { orientation: landscape; }",
      "errored": false,
      "fixed": false,
      "output": "@viewport { orientation: landscape; }",
      "reportedDisables": [],
      "results": [
        {
          "deprecations": [],
          "errored": false,
          "ignored": undefined,
          "invalidOptionWarnings": [],
          "parseErrors": [],
          "warnings": [],
        },
      ],
      "ruleMetadata": {
        "at-rule-no-vendor-prefix": {
          "fixable": true,
          "url": "https://stylelint.io/user-guide/rules/at-rule-no-vendor-prefix",
        },
      },
    },
  ],
}
`;
